hypervisor/src/mshv/mod.rs

```diff
@@ -52,6 +51,205 @@ pub struct HvState {
+                err
+            );
+        });
+
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); 2];
+
+        'epoll: loop {
+            let num_events = match epoll::wait(epoll_file.as_raw_fd(), -1, &mut events[..]) {
+                Ok(res) => res,
+                Err(e) => {
+                    if e.kind() == std::io::ErrorKind::Interrupted {
+                        continue;
+                    }
+                    panic!("irqfd epoll ???");
+                }
+            };
+
+            for event in events.iter().take(num_events) {
+                let ev_type = event.data as u16;
+
+                match ev_type {
+                    KILL_EVENT => {
+                        break 'epoll;
+                    }
+                    IRQFD_EVENT => {
+                        debug!("IRQFD_EVENT received, inject to guest");
+                        let _ = self.irqfd.read().unwrap();
+                        let gsi_routes = self.gsi_routes.read().unwrap();
+
+                        if let Some(e) = gsi_routes.get(&self.gsi) {
+                            assert_virtual_interrupt(&self.vm, &e);
+                        } else {
+                            debug!("No routing info found for GSI {}", self.gsi);
+                        }
+                    }
+                    _ => {
+                        error!("Unknown event");
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Translate from architectural defined delivery mode to Hyper-V type
+// See Intel SDM vol3 10.11.2
+fn get_interrupt_type(delivery_mode: u8) -> Option<hv_interrupt_type> {
+    match delivery_mode {
+        0 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_FIXED),
+        1 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_LOWESTPRIORITY),
+        2 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_SMI),
+        4 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_NMI),
+        5 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_INIT),
+        7 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_EXTINT),
+        _ => None,
+    }
+}
+
+// See Intel SDM vol3 10.11.1
+// We assume APIC ID and Hyper-V Vcpu ID are the same value
+// This holds true for HvLite
+fn get_destination(message_address: u32) -> u64 {
+    ((message_address >> 12) & 0xff).into()
+}
+
+fn get_destination_mode(message_address: u32) -> bool {
+    if (message_address >> 2) & 0x1 == 0x1 {
+        return true;
+    }
+
+    false
+}
+
+fn get_redirection_hint(message_address: u32) -> bool {
+    if (message_address >> 3) & 0x1 == 0x1 {
+        return true;
+    }
+
+    false
+}
+
+fn get_vector(message_data: u32) -> u8 {
+    (message_data & 0xff) as u8
+}
+
+// True means level triggered
+fn get_trigger_mode(message_data: u32) -> bool {
+    if (message_data >> 15) & 0x1 == 0x1 {
+        return true;
+    }
+
+    false
+}
+
+fn get_delivery_mode(message_data: u32) -> u8 {
+    ((message_data & 0x700) >> 8) as u8
+}
+
```

Comment:
Maybe it would be cleaner to make those `MshvIrqRoutingMsi` methods instead of having them floating around the code?

Reply:
This implementation mainly contributed by @liuw >I would let him comment.

Reply:
Yes. Making them methods of `MshvIrqRoutingMsi` is fine. They are not tied to MSHV in anyway, but making them methods won't hurt either.